th-249: Redesign UI table

h0wter opened the pull request:
Changed the table color scheme and text style.
Changed styles for pagination.
The table header and button from the design belong to other components and will be styled separately.
![table](https://github.com/BinaryStudioAcademy/bsa-2023-towhub/assets/91832203/17bbf5e5-0204-456c-9958-d658ea3a8cf4)

h0wter committed:
th-249: * update table styles

h0wter committed:
th-249: * update pagination and dropdown styles

h0wter committed:
Merge branch 'development' into task/th-249-redisign-ui-table

kalinkaaaa14 commented:
will this table have sorting on columns?

kalinkaaaa14 commented on frontend/src/libs/components/pagination/pagination.tsx:
```diff
@@ -143,6 +164,7 @@ const Pagination: React.FC<Properties> = ({
           <Dropdown
             options={createOptions([5, 10, 20, 50, 100])}
             defaultValue={createOption(pageSize)}
+            isValueCustomized
```
this value will be always true in dropdown, was it supposed to be?

h0wter replied:
The dropdown is styled so that the text of the selected option matches the filter styles on the Home Page. Here in the pagination, the styles are different and I can't override them with a passed class because the dropdown styles use nested selectors and the specificity is higher. I added this props to the pagination to remove the styling of the selected option and it inherits it from the pagination. 
In pagination this props will always be true, but in the rest of the application it will not be.

kalinkaaaa14 commented on frontend/src/libs/components/pagination/pagination.tsx:
```diff
@@ -122,17 +119,41 @@ const Pagination: React.FC<Properties> = ({
           isDisabled={isFirstPage}
         />
-        {isHiddenFirstPage ? <div className={styles.dots}>...</div> : null}
+        {isHiddenFirstPage ? (
+          <>
+            <Button
+              className={styles.btn}
+              label={'1'}
+              variant="text"
+              onClick={handlePageClick}
+            ></Button>
+            <div className={styles.dots}>...</div>
+          </>
+        ) : null}
```
```suggestion
        {isHiddenFirstPage && 
          <>
            <Button
              className={styles.btn}
              label='1'
              variant="text"
              onClick={handlePageClick}
            ></Button>
            <div className={styles.dots}>...</div>
          </>
        }
```

kalinkaaaa14 commented on frontend/src/libs/components/pagination/pagination.tsx:
```diff
@@ -122,17 +119,41 @@ const Pagination: React.FC<Properties> = ({
+              onClick={handlePageClick}
+            ></Button>
+          </>
+        ) : null}
```
```suggestion
        {isHiddenLastPage && 
          <>
            <div className={styles.dots}>...</div>
            <Button
              className={styles.btn}
              label={`${pageCount}`}
              variant="text"
              onClick={handlePageClick}
            ></Button>
          </>
         }
```

kalinkaaaa14 commented on frontend/src/libs/components/dropdown/dropdown.tsx:
```diff
@@ -28,18 +28,20 @@ type Properties<T extends FieldValues> = {
 
 const getClassNames = (
   isMenuOpen: boolean,
+  isValueCustomized: boolean,
```
```suggestion
  isCustomValueContainer: boolean,
```

h0wter committed:
th-249: + edit&delete icon and ability to sort columns

h0wter committed:
th-249: * changed props name

h0wter committed:
Merge branch 'development' into task/th-249-redisign-ui-table

h0wter committed:
th-24: * change imports

h0wter commented:
Added ability to edit/delete records from the table, for this you need to add `isTableEditable` props and `onEditClick` and `onDeleteClick` handlers.
Also added the ability to sort, for this you need to add `const [sorting, setSorting] = useState<SortingState>([])` to your component and pass to Table as props `sorting`, `setSorting`. To disallow sorting on a particular field, you need to add `enableSorting: false` to the column. You can also set a custom id for the column so that you can conveniently handle sorting from the state.

h0wter committed:
th-249: + add manualSorting parameter

Druxight commented on frontend/src/libs/components/pagination/pagination.tsx:
```diff
@@ -122,17 +119,41 @@ const Pagination: React.FC<Properties> = ({
+            <div className={styles.dots}>...</div>
+            <Button
+              className={styles.btn}
+              label={`${pageCount}`}
```
```suggestion
              label={pageCount}
```

Druxight commented on frontend/src/libs/components/pagination/pagination.tsx:
```diff
@@ -122,17 +119,41 @@ const Pagination: React.FC<Properties> = ({
+          <>
+            <Button
+              className={styles.btn}
+              label={'1'}
```
```suggestion
              label="1"
```

Druxight commented on frontend/src/libs/components/table/table.tsx:
```diff
@@ -90,37 +125,76 @@ const Table = <T,>({
+    index: number,
+    arrayLength: number,
+  ): typeof onEditClick | typeof onDeleteClick => {
+    if (isTableEditable && index === arrayLength - 2) {
```
it's not clear what magic numbers stand for.
move the index === arrayLength - 2 check into the separate constant with meaningful name

Druxight commented on frontend/src/libs/components/table/table.tsx:
```diff
@@ -90,37 +125,76 @@ const Table = <T,>({
+  ): typeof onEditClick | typeof onDeleteClick => {
+    if (isTableEditable && index === arrayLength - 2) {
+      return onEditClick;
+    } else if (isTableEditable && index === arrayLength - 1) {
```
the if-else block can be simplified

Druxight commented on frontend/src/libs/components/table/table.tsx:
```diff
@@ -90,37 +125,76 @@ const Table = <T,>({
+          <tr key={row.id} className={styles.tr}>
+            {row.getVisibleCells().map((cell, index, array) => {
+              const isEditIconCell =
+                isTableEditable && index === array.length - 2;
```
same check here, perhaps its better to move it into separate function

h0wter committed:
th-249: + add helper for checking icon cell

kalinkaaaa14 commented on frontend/src/libs/components/table/libs/helpers/check-icon-cell.helper.ts:
```diff
@@ -0,0 +1,16 @@
+const EDIT_ICON_POSITION_FROM_RIGHT = 2;
+const DELETE_ICON_POSITION_FROM_RIGHT = 1;
+
+type Return = {
```
```suggestion
type CheckIsIconCell = {
```

kalinkaaaa14 commented on frontend/src/libs/components/table/libs/helpers/check-icon-cell.helper.ts:
```diff
@@ -0,0 +1,16 @@
+  isDeleteCell: boolean;
+};
+
+const checkIconCell = (index: number, arrayLength: number): Return => {
```
```suggestion
const checIskIconCell = (index: number, arrayLength: number): Return => {
```

kalinkaaaa14 commented on frontend/src/libs/components/table/table.tsx:
```diff
@@ -90,37 +126,82 @@ const Table = <T,>({
+        return (
+          <tr key={row.id} className={styles.tr}>
+            {row.getVisibleCells().map((cell, index, array) => {
+              const isEditIconCell =
+                isTableEditable &&
+                checkIconCell(index, array.length).isEditCell;
+              const isDeleteIconCell =
+                isTableEditable &&
+                checkIconCell(index, array.length).isDeleteCell;
```
```suggestion
              const { isEditCell, isDeleteCell } =checkIsIconCell({isTableEditable, index, arrLength: array.length});
```

kalinkaaaa14 commented on frontend/src/libs/components/table/table.tsx:
```diff
@@ -90,37 +126,82 @@ const Table = <T,>({
+              const isDeleteIconCell =
+                isTableEditable &&
+                checkIconCell(index, array.length).isDeleteCell;
+              let icon = null;
+
+              if (isEditIconCell) {
+                icon = rowWithIcons.iconEdit;
+              } else if (isDeleteIconCell) {
+                icon = rowWithIcons.iconDelete;
+              }
+
```
remove this block

kalinkaaaa14 replied:
add const shouldAddIconToCell = isEditIconCell || isDeleteIconCell;

kalinkaaaa14 replied:
add const iconToDisplay = isEditIconCell ? IconName.EDIT : IconName.TRASH;

kalinkaaaa14 commented on frontend/src/libs/components/table/table.tsx:
```diff
@@ -90,37 +126,82 @@ const Table = <T,>({
+                  }}
+                  onClick={getHandler(index, array.length)?.bind(null, row.id)}
+                >
+                  {!icon &&
```
```suggestion
                  {!shouldAddIconToCell &&
```

kalinkaaaa14 commented on frontend/src/libs/components/table/table.tsx:
```diff
@@ -90,37 +126,82 @@ const Table = <T,>({
+                >
+                  {!icon &&
+                    flexRender(cell.column.columnDef.cell, cell.getContext())}
+                  {icon && <Icon iconName={icon} size="xl" />}
```
```suggestion
                  {shouldAddIconToCell && <Icon iconName={iconToDisplay } size="xl" />}
```

kalinkaaaa14 commented on frontend/src/libs/components/table/table.tsx:
```diff
@@ -90,37 +126,82 @@ const Table = <T,>({
-        </tr>
-      ))}
+      {table.getRowModel().rows.map((row) => {
+        const rowWithIcons = row.original as T & Icons;
```
I think you don't need to add icons, remove it
see the proposal below

kalinkaaaa14 commented on frontend/src/libs/components/table/libs/helpers/add-icons-to-data.helper.ts:
```diff
@@ -0,0 +1,44 @@
+): ReturnType<T> => {
+  const dataWithIcons = data.map((it) => ({
+    ...it,
+    iconEdit: IconName.EDIT,
+    iconDelete: IconName.TRASH,
```
I think you can remove it

h0wter committed:
th-249: * refactored table and helper

h0wter committed:
Merge branch 'development' into task/th-249-redisign-ui-table

kalinkaaaa14 commented on frontend/src/libs/components/table/table.tsx:
```diff
@@ -54,34 +78,42 @@ const Table = <T,>({
               key={header.id}
-              className={styles.th}
+              className={getValidClassNames(
+                styles.th,
```
let's make table/table.tsx more clear
create libs/helpers/helpers
libs/helpers/create-table-thead.helper
libs/helpers/create-table-tbody.helper
libs/helpers/get-cell-click-handler.helper

kalinkaaaa14 commented on frontend/src/libs/components/table/table.tsx:
```diff
@@ -90,37 +122,75 @@ const Table = <T,>({
+    if (!isTableEditable) {
+      return;
+    }
+    const { isEditCell } = checkIsIconCell(isTableEditable, index, arrayLength);
```
maybe let's better rename checkIsActionCell? for me sounds better, and don't forget about renaming the file

h0wter committed:
th-249: * splits table.tsx into different files

h0wter committed:
Merge branch 'development' into task/th-249-redisign-ui-table

h0wter committed:
Merge branch 'task/th-249-redisign-ui-table' of https://github.com/h0wter/bsa-2023-towhub into task/th-249-redisign-ui-table

h0wter committed:
th-249: - removes unused props

kalinkaaaa14 commented on frontend/src/libs/components/table/libs/components/thead.tsx:
```diff
@@ -0,0 +1,55 @@
+  table: Table<T>;
+};
+
+const Thead = <T,>({ table }: Properties<T>): JSX.Element => (
```
```suggestion
const Thead = <T>({ table }: Properties<T>): JSX.Element => (
```

PavloSukhinin commented on frontend/src/libs/components/dropdown/dropdown.tsx:
```diff
@@ -28,18 +28,20 @@ type Properties<T extends FieldValues> = {
+  isCustomValueContainer?: boolean;
 };
 
 const getClassNames = (
   isMenuOpen: boolean,
+  isCustomValueContainer: boolean,
 ): ClassNamesConfig<SelectOption, false, GroupBase<SelectOption>> => ({
```
```suggestion
type FooArgs = {
  isMenuOpen: boolean;
  isCustomValueContainer: boolean;
}

const getClassNames = ({
isMenuOpen,
isCustomValueContainer
}: FooArgs): ClassNamesConfig<SelectOption, false, GroupBase<SelectOption>> => ({
```
because many positional arguments (especially booleans) is ambiguous

kalinkaaaa14 commented on frontend/src/libs/components/dropdown/dropdown.tsx:
```diff
@@ -28,18 +28,20 @@ type Properties<T extends FieldValues> = {
   placeholder: () => styles.placeholder,
   singleValue: () => styles.singleValue,
-  valueContainer: () => styles.valueContainer,
+  valueContainer: () => (isCustomValueContainer ? '' : styles.valueContainer),
```
I don't like the idea with empty string, let's rewrite it ?

```
const getClassNames = (
  isMenuOpen: boolean,
  isCustomValueContainer: boolean,
): ClassNamesConfig<SelectOption, false, GroupBase<SelectOption>> => {
  const commonStylesConfig: ClassNamesConfig<
    SelectOption,
    false,
    GroupBase<SelectOption>
  > = {
    container: () => styles.container,
    control: () => styles.control,
    option: () => styles.option,
    menu: () => styles.singleValue,
    placeholder: () => styles.placeholder,
    singleValue: () => styles.singleValue,
    dropdownIndicator: () =>
      isMenuOpen
        ? getValidClassNames(styles.dropdownIndicator, styles.upside)
        : styles.dropdownIndicator,
    indicatorSeparator: () => styles.indicatorSeparator,
  };

  return isCustomValueContainer
    ? commonStylesConfig
    : { ...commonStylesConfig, valueContainer: () => styles.valueContainer };
};
```

PavloSukhinin commented on frontend/src/libs/components/table/libs/helpers/check-is-action-cell.helper.ts:
```diff
@@ -0,0 +1,23 @@
+  isDeleteCell: boolean;
+};
+
+const checkIsActionCell = (
+  isTableEditable: boolean,
+  index: number,
+  arrayLength: number,
+): CheckIsIconCell => {
```
```suggestion
const checkIsActionCell = ({
  isTableEditable,
  index,
  arrayLength, // and let's rename this one
}): CheckIsIconCell => {
```

PavloSukhinin commented on frontend/src/libs/components/table/libs/helpers/get-cell-click-handler.helper.ts:
```diff
@@ -0,0 +1,30 @@
+
+const getCellClickHandler = ({
+  index,
+  arrayLength,
```
Can we rename this parameter?

PavloSukhinin commented on frontend/src/libs/components/table/libs/helpers/get-cell-click-handler.helper.ts:
```diff
@@ -0,0 +1,30 @@
+import { checkIsActionCell } from './check-is-action-cell.helper.js';
+
+type Parameters = {
```
It's built-in type [Parameters](https://www.typescriptlang.org/docs/handbook/utility-types.html#parameterstype) 🥲

PavloSukhinin commented on frontend/src/libs/components/table/libs/helpers/add-icons-to-data.helper.ts:
```diff
@@ -0,0 +1,33 @@
+import { type ColumnDef } from '~/libs/types/types.js';
+
+type addIconsToData<T> = {
```
capitalize type name

h0wter committed:
th-249: * changed types and types' names

h0wter committed:
Merge branch 'development' into task/th-249-redisign-ui-table
